src/main/drivers/serial_uart_stm32f4xx.c

```diff
@@ -42,182 +45,231 @@ typedef enum UARTDevice {
-    .rcc_uart = RCC_APB1(USART3),
-    .rxIrq = USART3_IRQn,
-    .txPriority = NVIC_PRIO_SERIALUART3_TXDMA,
-    .rxPriority = NVIC_PRIO_SERIALUART3
-};
+# define UART3_TX_DMA_STREAM DMA1_Stream3
+#else
+# define UART3_TX_DMA_STREAM 0
 #endif
 
-#ifdef USE_UART4
-static uartDevice_t uart4 =
-{
-    .DMAChannel = DMA_Channel_4,
 #ifdef USE_UART4_RX_DMA
-    .rxDMAStream = DMA1_Stream2,
+# define UART4_RX_DMA_STREAM DMA1_Stream2
+#else
+# define UART4_RX_DMA_STREAM 0
 #endif
 #ifdef USE_UART4_TX_DMA
-    .txDMAStream = DMA1_Stream4,
-#endif
-    .dev = UART4,
-    .rx = IO_TAG(UART4_RX_PIN),
-    .tx = IO_TAG(UART4_TX_PIN),
-    .af = GPIO_AF_UART4,
-    .rcc_uart = RCC_APB1(UART4),
-    .rxIrq = UART4_IRQn,
-    .txPriority = NVIC_PRIO_SERIALUART4_TXDMA,
-    .rxPriority = NVIC_PRIO_SERIALUART4
-};
+# define UART4_TX_DMA_STREAM DMA1_Stream4
+#else
+# define UART4_TX_DMA_STREAM 0
 #endif
 
-#ifdef USE_UART5
-static uartDevice_t uart5 =
-{
-    .DMAChannel = DMA_Channel_4,
 #ifdef USE_UART5_RX_DMA
-    .rxDMAStream = DMA1_Stream0,
+# define UART5_RX_DMA_STREAM DMA1_Stream0
+#else
+# define UART5_RX_DMA_STREAM 0
 #endif
 #ifdef USE_UART5_TX_DMA
-    .txDMAStream = DMA1_Stream7,
-#endif
-    .dev = UART5,
-    .rx = IO_TAG(UART5_RX_PIN),
-    .tx = IO_TAG(UART5_TX_PIN),
-    .af = GPIO_AF_UART5,
-    .rcc_uart = RCC_APB1(UART5),
-    .rxIrq = UART5_IRQn,
-    .txPriority = NVIC_PRIO_SERIALUART5_TXDMA,
-    .rxPriority = NVIC_PRIO_SERIALUART5
-};
+# define UART5_TX_DMA_STREAM DMA1_Stream7
+#else
+# define UART5_TX_DMA_STREAM 0
 #endif
 
-#ifdef USE_UART6
-static uartDevice_t uart6 =
-{
-    .DMAChannel = DMA_Channel_5,
 #ifdef USE_UART6_RX_DMA
-    .rxDMAStream = DMA2_Stream1,
-#endif
-#ifdef USE_UART6_TX_DMA
-    .txDMAStream = DMA2_Stream6,
-#endif
-    .dev = USART6,
-    .rx = IO_TAG(UART6_RX_PIN),
-    .tx = IO_TAG(UART6_TX_PIN),
-    .af = GPIO_AF_USART6,
-    .rcc_uart = RCC_APB2(USART6),
-    .rxIrq = USART6_IRQn,
-    .txPriority = NVIC_PRIO_SERIALUART6_TXDMA,
-    .rxPriority = NVIC_PRIO_SERIALUART6
-};
-#endif
-
-static uartDevice_t* uartHardwareMap[] = {
-#ifdef USE_UART1
-    &uart1,
-#else
-    NULL,
-#endif
-#ifdef USE_UART2
-    &uart2,
+# define UART6_RX_DMA_STREAM DMA2_Stream1
 #else
-    NULL,
+# define UART6_RX_DMA_STREAM 0
 #endif
-#ifdef USE_UART3
-    &uart3,
-#else
-    NULL,
-#endif
-#ifdef USE_UART4
-    &uart4,
-#else
-    NULL,
-#endif
-#ifdef USE_UART5
-    &uart5,
+#ifdef USE_UART6_TX_DMA
+# define UART6_TX_DMA_STREAM DMA2_Stream6
 #else
-    NULL,
+# define UART6_TX_DMA_STREAM 0
 #endif
-#ifdef USE_UART6
-    &uart6,
+
+static uartDevice_t uartHardware[] = {
+    {
+        .dev = USART1,
+        .DMAChannel = DMA_Channel_4,
+        .rxDMAStream = UART1_RX_DMA_STREAM,
+        .txDMAStream = UART1_TX_DMA_STREAM,
+        .pinPair = {
+            { IO_TAG(PA10), IO_TAG(PA9) },
+            { IO_TAG(PB7), IO_TAG(PB6) },
+            { IO_TAG_NONE, IO_TAG_NONE }
+        },
+        .rx = IO_TAG_NONE,
```

Comment:
Just omitting it in initializer will set it to NONE - unset fields are zeroed, zeroed ioTag/IO_t is by design equal to NONE...